Fsync

fcami opened the pull request:
We do not fclose() before rename() on successful completion of writeConfig.

This is evident using strace:
rename("../boot/grub2/grub.cfg-", "../boot/grub2/grub.cfg") = 0
write(3, "ource $prefix/custom.cfg;\nfi\n### END /etc/grub.d/41_custom ###\n", 63) = 63
exit_group(0)      

We do not fsync() the configuration file, nor the directory, which can result in a boot failure on powerloss because the configuration file is truncated in unpredictable ways.

This branch aims to fix this and achieve an atomic rewrite of grub2's configuration file.

Resulting strace output:
write(3, "ource $prefix/custom.cfg;\nfi\n### END /etc/grub.d/41_custom ###\n", 63) = 63
fsync(3)                                = 0
close(3)                                = 0
munmap(0x7fdadba2f000, 4096)            = 0
rename("../boot/grub2/grub.cfg-", "../boot/grub2/grub.cfg") = 0
open("../boot/grub2", O_RDONLY)         = 3
fsync(3)                                = 0
close(3)                                = 0
exit_group(0)                           = ?

rhbz:
https://bugzilla.redhat.com/show_bug.cgi?id=1464611
https://bugzilla.redhat.com/show_bug.cgi?id=1466036

nixpanic commented on grubby.c:
```diff
@@ -1913,6 +1913,14 @@ static int writeConfig(struct grubConfig *cfg, char *outName,
                         return 1;
                 }
 
+		int outtempfd = open(tmpOutName, O_RDONLY);
```
This looks really weird, and I dont think it'll result in what you're hoping for (two different, unrelated file descriptors). To call `fsync()` on the `FILE*` you can obtain a file descriptor with `fileno()`. Doing this before the `fclose()` above makes it better understandable.

nixpanic commented on grubby.c:
```diff
@@ -1918,6 +1920,23 @@ static int writeConfig(struct grubConfig *cfg, char *outName,
+		int outfd = open(outName, O_RDONLY);
+		// open both dirname(outName) and dirname(tmpOutName) and pass both fds to fsync()
+		int dirfd1 = open(dirname(strdupa(outName)), O_RDONLY);
+		int dirfd2 = open(dirname(strdupa(tmpOutName)), O_RDONLY);
```
In case both files are in the same directory, you can skip one set of the `open()` and `fsync()` calls.

nixpanic commented on grubby.c:
```diff
@@ -1903,6 +1903,8 @@ static int writeConfig(struct grubConfig *cfg, char *outName,
 
 	if (tmpOutName) {
+		// fclose typically includes fflush but not fsync
+		// so this is not enough
```
See the note about `fileno()` in the next commit.

fcami commented:
I rebased and squashed the pointless commits - as it stands it's a single bug and does not warrant multiple commits I assume. Please let me know what you think.

vathpela commented:
Pulled to master as https://github.com/rhboot/grubby/commit/d9406b061cef40bd42ed813e13bd02b3a285adae
